### PR TITLE
WIP fix for copy as to do with wrong url

### PR DESCRIPTION
**Pull request by Feywood**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)
I've added in a function to return a different formatted note depending on group type being party/guild/tavern, but I'm having trouble inserting it into that notes constant.
Popping in the function call breaks the copy as to do button and I can't concatenate it with other text.
My gut is telling me I should be working async, but I don't have much of a handle on that, so any feedback is welcome.

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
WIP for https://github.com/HabitRPG/habitica/issues/10016

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
Have created function to return correct message format

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 9c4776eb-7330-465b-85cc-e4ccef6c1ad1

### CODE REVIEW

File: website/client/components/chat/copyAsTodoModal.vue [hunk 1]
```diff
@@ -36,9 +36,18 @@ export default {
...
+    } else if (this.group.Type === 'guild') {
+      return `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}]([${this.groupName}](${baseURL}/groups/guild/${this.groupId})`;
+    } else if (this.groupId === 'tavern') {
+      return `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}](${baseURL}/groups/tavern)`;
+    }
+  },
   mounted () {
     this.$root.$on('habitica::copy-as-todo', message => {
-      const notes = `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}](${baseUrl}/groups/guild/${this.groupId})`;
+      const notes = generateURL();
```
Comment (negue) [inline website/client/components/chat/copyAsTodoModal.vue]: this should be `this.generateURL(message)`

File: website/client/components/chat/copyAsTodoModal.vue [hunk 2]
```diff
@@ -36,9 +36,18 @@ export default {
       task: {},
     };
   },
+  generateURL () {
```
Comment (negue) [inline website/client/components/chat/copyAsTodoModal.vue]: you should add `message` to `generateURL` or else you won't have a message to use ^^

File: website/client/components/chat/copyAsTodoModal.vue [hunk 3]
```diff
@@ -36,9 +36,18 @@ export default {
...
+    if (this.group.Type === 'party') {
+      return `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}](${baseURL}/party)`;
+    } else if (this.group.Type === 'guild') {
+      return `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}]([${this.groupName}](${baseURL}/groups/guild/${this.groupId})`;
+    } else if (this.groupId === 'tavern') {
+      return `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}](${baseURL}/groups/tavern)`;
```
Comment (negue) [inline website/client/components/chat/copyAsTodoModal.vue]: `$baseURL` needs to be `$baseUrl`

File: website/client/components/chat/copyAsTodoModal.vue [hunk 4]
```diff
@@ -36,9 +36,18 @@ export default {
       task: {},
     };
   },
+  generateURL () {
+    if (this.group.Type === 'party') {
+      return `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}](${baseURL}/party)`;
```
Comment (negue) [inline website/client/components/chat/copyAsTodoModal.vue]: `${message.user || 'system message'}${message.user ? ' wrote' : ''} in [${this.groupName}]` is three times in one method, I'd suggest to extract it to a local variable and then use that variable for each of the urls

**Review by negue — CHANGES_REQUESTED** [review]
the direction is correct, just some issues in the code. 

Also there is no need to get `async` on this change, its just a string concat

**Comment by Feywood** [comment]
Thanks so much for your suggestions, it did clarify a few things for me.
However, it's not working. Either I misunderstood what you meant or it's just...
In the current state, my console in chrome tells me generateURL is not a function. Defining it as a function with function:{} did not change this, nor did adding it onto the methods (this just got me can't resolve type of undefined).
So clearly I'm still missing something somewhere.
I'm learning, I guess?

**Comment by paglias** [comment]
@Feywood generateUrl is a Vue method, so you can't access it from the global scope. When you do `console.log(functionName)` it's actually doing `window.functionName` but in this case the function is registered under the vue component, not the window (global scope). To log it you can do `console.log(this.functionName)` here https://github.com/HabitRPG/habitica/pull/10090/files#diff-59ae0c35bb3ef4d17ed8cb499eaee8cfL41 inside `mounted()`, `this` means you're referring to the Vue component scope

File: website/client/components/chat/copyAsTodoModal.vue [hunk 5]
```diff
@@ -36,9 +36,19 @@ export default {
       task: {},
     };
   },
+  generateURL (message) {
```
Comment (negue) [inline website/client/components/chat/copyAsTodoModal.vue]: this needs to be placed inside `methods: { }`, or else it won't be recognized as a vue method

**Comment by lemoness** [comment]
@Feywood - Hi there! How is this fix going?

**Comment by Feywood** [comment]
@lemoness I'm still blocked despite the assistance above. Will post a more substantial update about my issue in next couple days.

**Comment by beffymaroo** [comment]
@Feywood How's it going? Are you still blocked? Let us know when you have a chance. Thank you!

**Comment by Feywood** [comment]
Having finally opened my code again: yes, moving the method into methods has it recognised as such, but now there is a new scope issue in that the group object the logic is based on is considered undefined. 
Not knowing vue well enough, I'm not sure how to move it between the two scopes as simply adding it into the method variables does not seem to be  doing the trick. @negue if you have any pointers, I would very much welcome them.

**Comment by negue** [comment]
@Feywood either you add a new `prop` like `groupType` which you could use instead of `this.group.Type` or switch from `groupName` and `groupId` to a property which contains the full group, but then you'd have to change all occurrences

**Comment by Feywood** [comment]
Alright, there is actual progress this time.
Having found the existing props groupId and groupName, the code is now functional for the tavern, hurrah hurrah.
However (as always), not yet for party/guild. There the issue is that of course groupNames are entirely up to users so I can't use that variable, and groupId is changable as well, so not so much either. I cannot seem to derive a group type from ID or name, so no luck.
The other option seems to be generating a prop myself with the grouptype, but while I can see the actual prop in my dev console, I cannot seem to get it to populate and it remains 'undefined'. I'm guessing this info has to come from chatMessages.vue or group.vue but how?

**Comment by negue** [comment]
In your recent commit you removed the support for party & group messages, this was a step backwards^^

the issue was, like I wrote in my prior message, you either needed a new `prop`erty for `groupType` or pass the full `group` , in the code the `this.group` was undefined

For debugging steps is always helpful just to use `console.info`, just put all variables in it and then you'll see quite fast which one is `undefined`

**Comment by beffymaroo** [comment]
@Feywood How's this fix going? Let us know if you're still working on it when you get a chance. Thank you!

**Pull request closed without merging**